README.md

```diff
@@ -1,16 +1,658 @@
+
+The OPTIONAL `run` field MUST reference the actions contained in the UCAN are to be run during the invocation. To run all actions in the underlying UCAN, the `"*"` value MUST be used. If only specific actions (or [pipelines](#5-promise-pipelining)) are intended to be run, then they MUST be prefixed with an arbitrary label and treated as a UCAN attenuation: all actions MUST be backed by a matching capability of equal or lesser scope.
+
+#### 3.1.3.1 Promises
+
+The only difference from general capabilities is that [promises](#5-promise-pipelining) MAY also be used as inputs to attenuated fields.
+
+If a capability input has the key `"_"` and the value is a promise, the input MUST be discarded and only used for determining sequencing actions.
+
+### 3.1.4 Nonce
+
+The `nnc` field MUS contain a unique nonce. This field makes each invocation unique.
+
+### 3.1.5 Extended Fields
+
+The OPTIONAL `ext` field MAY contain arbitrary data. If not present, the `ext` field MUST me interpreted as `null`, including for [signature](#315-signature).
+
+### 3.1.6 Signature
+
+The `sig` field MUST contain the signature of the other fields. To construct the payload, the other fields MUST first be canonicalized as `dag-cbor`.
+
+If serialized as JSON, the `sig` field MUST be serialized as [unpadded base64url](https://datatracker.ietf.org/doc/html/rfc4648#section-5).
+
+## 3.2 IPLD Schema
+
+``` ipldsch
+type Invocation struct {
+  inv &UCAN  -- The UCAN providing authority
+  v   SemVer -- Version
+  run Scope  -- Which actions to invoke
+  nnc String -- Nonce
+  ext Any    -- Extended fields
+  sig Bytes  -- Signature
+}
+
+type Scope enum {
+  | All ("*")
+  | Action
+}
+
+type Action enum {
+  using  URI 
+  do     Ability
+  inputs Any
+} 
+```
+
+## 3.3 JSON Examples
+
+### 3.3.1 Run All
+
+ ``` js
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "ext": null,
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+ ``` js
+{
+  "ucan/invoke": "QmYW8Z58V1v8R25USVPUuFHtU7nGouApdGTk3vRPXmVHPR",
+  "v": "0.1.0",
+  "nnc": "abcdef",
+  "run": "*", // Explicitly "run all"
+  "ext": null,
+  "sig": "bdNVZn_uTrQ8bgq5LocO2y3gqIyuEtvYWRUH9YT-SRK6v_SX8bjt-VZ9JIPVTdxkWb6nhVKBt6JGpgnjABpOCA"
+}
+```
+
+### 3.3.2 Promise Pipelines
+
+Promise pipelines are handled in more detail in [section 5](#5-promise-pipelining). In brief, they enable taking the output of an action that has not yet run as the input to another action. Here is a simple example:
+
+``` js
+{
+  "ucan/invoke": "QmY4jEVE35u8SHegWoDak2x6vTAZ6Cc4cpSD5LAUQ23W7L",
+  "v": "0.1.0",
+  "nnc": "02468",
+  "ext": null,
+  "run": {
+    "notify-bob": {
+      "using": "mailto://alice@example.com",
+      "do": "msg/send",
+      "inputs": [
+        {
+          "to": "bob@example.com",
+          "subject": "DNSLink for example.com",
+          "body": "Hello Bob!"
+        }
+      ]
+    },
```

Comment:
I assume these blocks reference things in the `ucan/invoke`? Can we see an example of what's in that object?

Reply:
@simonwo In they current design, they don't because we know the types from the wrapper. I have a few examples over in IPVM, which I'm cleaning up today. I see that you have comments there, I'll respond in the IPVM repo.